Codec/Archive/Tar/Unpack.hs

```diff
@@ -83,32 +85,59 @@ unpack baseDir entries = do
     -- We're relying here on 'checkSecurity' to make sure we're not scribbling
     -- files all over the place.
 
-    unpackEntries _     (Fail err)      = either throwIO throwIO err
-    unpackEntries links Done            = return links
-    unpackEntries links (Next entry es) = case entryContent entry of
-      NormalFile file _ -> do
-        extractFile (entryPermissions entry) (entryPath entry) file (entryTime entry)
-        unpackEntries links es
-      Directory         -> do
-        extractDir (entryPath entry) (entryTime entry)
-        unpackEntries links es
-      HardLink     link -> (unpackEntries $! saveLink True (entryPath entry) link links) es
-      SymbolicLink link -> (unpackEntries $! saveLink False (entryPath entry) link links) es
-      OtherEntryType 'L' (Char8.unpack . BS.toStrict -> fn) _ ->
-        case es of
-             (Next entry' es') -> case entryContent entry' of
-               NormalFile file _ -> do
-                extractFile (entryPermissions entry') fn file (entryTime entry')
-                unpackEntries links es'
-               Directory         -> extractDir fn (entryTime entry')
-                                 >> unpackEntries links es'
-               HardLink     link -> (unpackEntries $! saveLink True fn link links) es'
-               SymbolicLink link -> (unpackEntries $! saveLink False fn link links) es'
-               OtherEntryType 'L' _ _ -> throwIO $ userError "Two subsequent OtherEntryType 'L'"
-               _ -> unpackEntries links es'
-             (Fail err)      -> either throwIO throwIO err
-             Done            -> throwIO $ userError "././@LongLink without a subsequent entry"
-      _ -> unpackEntries links es --ignore other file types
+    unpackEntries :: Exception e
+                  => Maybe LinkTarget
+                  -> Maybe FilePath
+                  -> [(FilePath, FilePath, Bool)]     -- ^ links (path, link, isHardLink)
+                  -> Entries (Either e FileNameError) -- ^ entries
+                  -> IO [(FilePath, FilePath, Bool)]
+    unpackEntries _ _ _     (Fail err)      = either throwIO throwIO err
+    unpackEntries _ _ links Done            = return links
+    unpackEntries mLink mPath links (Next entry es) = do
```

Comment:
The interesting case is when we have `OtherEntryType 'K'` followed immediately by `OtherEntryType 'L'`, which means we have a symbolic link, where:

* the symlink target is too long
* the destination filepath is too long as well

Our implementation here is a little more permissive as it allows the `[ OtherEntryType 'L', OtherEntryType 'K']` as well. I think the spec is not very clear on whether that's valid.